internal/stage_h5.go

```diff
@@ -0,0 +1,163 @@
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "$ echo " + randomWords3,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^\s*echo ` + randomWords3 + `\s*$`),
+		},
+		StayOnSameLine: true,
+	})
+	if err := asserter.AssertWithoutPrompt(); err != nil {
+		return err
+	}
+	asserter.PopAssertion()
+	stageHarness.Logger.Successf("✓ Prompt line matches %q", "echo "+randomWords3)
+
+	// Second up-arrow should recall the invalid command
+	if err := shell.SendCommandRaw(upArrow); err != nil {
+		return err
+	}
+	stageHarness.Logger.Infof("Pressed %q (expecting to recall %q)", "<UP ARROW>", randomCommand)
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "$ " + randomCommand,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^\s*` + randomCommand + `\s*$`),
+		},
+		StayOnSameLine: true,
+	})
+	if err := asserter.AssertWithoutPrompt(); err != nil {
+		return err
+	}
+	asserter.PopAssertion()
+	stageHarness.Logger.Successf("✓ Prompt line matches %q", randomCommand)
+
+	// Third up-arrow should recall echo randomWords2
+	if err := shell.SendCommandRaw(upArrow); err != nil {
+		return err
+	}
+	stageHarness.Logger.Infof("Pressed %q (expecting to recall %q)", "<UP ARROW>", "echo "+randomWords2)
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "$ echo " + randomWords2,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^\s*echo ` + randomWords2 + `\s*$`),
+		},
+		StayOnSameLine: true,
+	})
+	if err := asserter.AssertWithoutPrompt(); err != nil {
+		return err
+	}
+	asserter.PopAssertion()
+	stageHarness.Logger.Successf("✓ Prompt line matches %q", "echo "+randomWords2)
+
+	// Down-arrow should go forward to the invalid command
+	if err := shell.SendCommandRaw(downArrow); err != nil {
+		return err
+	}
+	stageHarness.Logger.Infof("Pressed %q (expecting to recall %q)", "<DOWN ARROW>", randomCommand)
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "$ " + randomCommand,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^\s*` + randomCommand + `\s*$`),
+		},
+		StayOnSameLine: true,
+	})
+	if err := asserter.AssertWithoutPrompt(); err != nil {
+		return err
+	}
+	asserter.PopAssertion()
+	stageHarness.Logger.Successf("✓ Prompt line matches %q", randomCommand)
+
+	// Execute the invalid command again
+	if err := shell.SendCommandRaw("\n"); err != nil {
+		return err
+	}
+	stageHarness.Logger.Infof("Executing command %q", randomCommand)
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "$ " + randomCommand,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^\s*` + randomCommand + `\s*$`),
+		},
+	})
+	asserter.AddAssertion(assertions.SingleLineAssertion{
+		ExpectedOutput: "command not found: " + randomCommand,
+		FallbackPatterns: []*regexp.Regexp{
+			regexp.MustCompile(`^.*command not found.*` + randomCommand + `.*$`),
```

Comment:
That's a lot of duplication from `InvalidCommandTestCase` - try instead changing this so that we're running one of the `echo` commands and not an invalid command? That will be simpler to assert for.